### PR TITLE
Create and enable a snmp trap handling container

### PR DESCRIPTION
**Pull request by glance-**
This enables and configures a container for ingesting snmp traps into
librenms.

### CODE REVIEW

**Review by crazy-max — APPROVED** [review]
LGTM. PTAL @murrant

File: Dockerfile [hunk 1]
```diff
@@ -130,7 +130,7 @@ RUN apk --update --no-cache add -t build-dependencies \
 
 COPY rootfs /
 
-EXPOSE 8000 514 514/udp
+EXPOSE 8000 514 514/udp 162 162/udp
```
Comment (systemcrash) [inline Dockerfile:133]: TCP port 162 is unnecessary. All SNMP transports are UDP.  Even DTLS is UDP - although that is port 10161/2 - can the sidecar handle encrypted traps?

File: examples/compose/docker-compose.yml [hunk 2]
```diff
@@ -138,3 +138,35 @@ services:
...
+    hostname: librenms-snmptrapd
+    cap_add:
+      - NET_ADMIN
+      - NET_RAW
+    depends_on:
+      - librenms
+    ports:
+      - target: 162
+        published: 162
+        protocol: tcp
```
Comment (systemcrash) [inline examples/compose/docker-compose.yml:154]: Remove....  SNMP is UDP only.

File: examples/rrdcached-server/docker-compose.yml [hunk 3]
```diff
@@ -156,3 +156,35 @@ services:
...
+    hostname: librenms-snmptrapd
+    cap_add:
+      - NET_ADMIN
+      - NET_RAW
+    depends_on:
+      - librenms
+    ports:
+      - target: 162
+        published: 162
+        protocol: tcp
```
Comment (systemcrash) [inline examples/rrdcached-server/docker-compose.yml:172]: Remove.... SNMP is UDP only.

File: examples/traefik/docker-compose.yml [hunk 4]
```diff
@@ -167,3 +167,32 @@ services:
...
+  snmptrapd:
+    image: librenms/librenms:latest
+    container_name: librenms_snmptrapd
+    hostname: librenms-snmptrapd
+    depends_on:
+      - librenms
+    ports:
+      - target: 162
+        published: 162
+        protocol: tcp
```
Comment (systemcrash) [inline examples/traefik/docker-compose.yml:180]: Remove.... SNMP is UDP only.

File: doc/notes/snmptrapd.md [hunk 5]
```diff
@@ -0,0 +1,12 @@
+## Snmptrapd
+
+If you want to enable snmptrapd, you have to run a "sidecar" container (see snmptrapd service in [docker-compose.yml](../../examples/compose/docker-compose.yml) example) or run a simple container like this :
+
+```bash
+docker run -d --name librenms_snmptrapd \
+  --env-file $(pwd)/librenms.env \
+  -e SIDECAR_SNMPTRAPD=1 \
+  -p 162 -p 162/udp \
```
Comment (systemcrash) [inline doc/notes/snmptrapd.md:9]: Remove `-p 162` .... SNMP is UDP only.

Might be an idea to change to 10161+10162 for DTLS encrypted traps.

**Review by systemcrash — COMMENTED** [review]
TCP port 162 is unnecessary as a privileged port (traps go via UDP). Coopt the port for 10161 and 10162 for DTLS instead.

**Comment by systemcrash** [comment]
Would be nice to have SNMP as a first class citizen in containers.

**Comment by glance-** [comment]
Not all equipment can send traps to another port than 161, so its a bad idea to pick something else by default. If you don't want to use port 161 in your setup, thats up to you.

And, snmp isn't udp only. Not even snmp traps is udp only, so no, tcp bits should be tere.

**Comment by crazy-max** [comment]
@systemcrash PTAL of @murrant would be nice. I'm not quite experienced with SNMP traps.

I think this PR is the best approach so far (yes there are currently 3 PRs to add SNMP trap support) but we could make things more configurable for this sidecar container as it is done in #169 and also allow a customize MIBs dir (see https://github.com/librenms/docker/pull/169#pullrequestreview-611747568).

File: rootfs/etc/snmp/snmptrapd.conf [hunk 6]
```diff
@@ -0,0 +1,3 @@
+disableAuthorization yes
+authCommunity log,execute,net COMMUNITYSTRING
```
Comment (chaitu-tk) [inline rootfs/etc/snmp/snmptrapd.conf:2]: It will be good to have credentials as runtime inputs to avoid rebuilding the image, maybe environment variables?

File: rootfs/etc/cont-init.d/08-svc-snmptrapd.sh [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+mkdir -p /data/snmptrapd /run/snmptrapd
+chown librenms. /data/snmptrapd
+chown -R librenms. /run/snmptrapd
+
+# Create service
+mkdir -p /etc/services.d/snmptrapd
+cat > /etc/services.d/snmptrapd/run <<EOL
+#!/usr/bin/execlineb -P
+with-contenv
+/usr/sbin/snmptrapd -f -m IF-MIB -M /opt/librenms/mibs:/opt/librenms/mibs/cisco
```
Comment (chaitu-tk) [inline rootfs/etc/cont-init.d/08-svc-snmptrapd.sh:23]: The `-m` option should be `-m +IF-MIB` without the `+` only `IF-MIB` is included and nothing else or we could simply do `-m ALL`

File: Dockerfile (hunk 1 above)
Comment (chaitu-tk) [reply Dockerfile:133]: we need to extend the snmptrapd.conf to add credentials for SNMPv2c (authCommunity) (already there but changes need docker image rebuild) and SNMPv3 (createUser) for both USM (Password-based) and TSM (DTLS - certs based), but as that is seldom used, we could start with simple v2c/v3-USM support and extend it later.

File: examples/traefik/docker-compose.yml [hunk 8]
```diff
@@ -167,3 +167,32 @@ services:
...
+
+  snmptrapd:
+    image: librenms/librenms:latest
+    container_name: librenms_snmptrapd
+    hostname: librenms-snmptrapd
+    depends_on:
+      - librenms
+    ports:
+      - target: 162
+        published: 162
```
Comment (chaitu-tk) [inline examples/traefik/docker-compose.yml:179]: I have tried this PR with some local changes, snmptraps are not seen in the LibreNMS, because of the docker networking.

By default docker uses `bridge` network mode and translated the source IP for these published ports (`DNAT`), so, `snmptrapd` container does not know which host has sent this (both hostname and IP address are translated), so, it just ignores the trap.

I have tried several combinations of networking but none of them works. 

E.g., using `host` mode networking does not allow containers to talk to each other
E.g. tried some custom networking options, but disabling source address translation for NAT doesn't make sense.

So, we need a networking scheme with the below properties, but there is no such mode in docker AFAIK (I am still looking at custom iptables for docker).

* Bridge all containers
* Published ports for `snmptrapd` container should act like `host` mode or at least no NATing.

And we cannot even workaround this problem as the source information is lost, only way I can see now is to run all of them in a single container.

**Review by chaitu-tk — CHANGES_REQUESTED** [review]
Docker networking is a blocker for this as we need the source IP of the SNMP notification.

File: examples/traefik/docker-compose.yml (hunk 8 above)
Comment (chaitu-tk) [reply examples/traefik/docker-compose.yml:179]: Well, I found a working way with `host` networking driver in docker with the below patch, earlier I was trying with `localhost` and MySQL was suing UNIX socket by default as they are in separate containers it was failing, I have modified to use IP address variant of `localhost` and it worked.

```
diff --git a/docker-compose.yml b/docker-compose.yml
index 6b210fc..8fb1828 100644
--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -4,6 +4,7 @@ services:
   db:
     image: mariadb:10.4
     container_name: librenms_db
+    network_mode: host
     command:
       - "mysqld"
       - "--innodb-file-per-table=1"
@@ -23,6 +24,7 @@ services:
   memcached:
     image: memcached:alpine
     container_name: librenms_memcached
+    network_mode: host
     environment:
       - "TZ=${TZ}"
     restart: always
@@ -30,6 +32,7 @@ services:
   redis:
     image: redis:5.0-alpine
     container_name: librenms_redis
+    network_mode: host
     environment:
       - "TZ=${TZ}"
     restart: always
@@ -37,6 +40,7 @@ services:
   msmtpd:
     image: crazymax/msmtpd:latest
     container_name: librenms_msmtpd
+    network_mode: host
     env_file:
       - "./msmtpd.env"
     restart: always
@@ -44,6 +48,7 @@ services:
   librenms:
     image: /:latest
     container_name: librenms
+    network_mode: host
     hostname: librenms
     cap_add:
       - NET_ADMIN
@@ -64,7 +69,7 @@ services:
       - "TZ=${TZ}"
       - "PUID=${PUID}"
       - "PGID=${PGID}"
-      - "DB_HOST=db"
+      - "DB_HOST=127.0.0.1"
       - "DB_NAME=${MYSQL_DATABASE}"
       - "DB_USER=${MYSQL_USER}"
       - "DB_PASSWORD=${MYSQL_PASSWORD}"
@@ -74,6 +79,7 @@ services:
   dispatcher:
     image: /:latest
     container_name: librenms_dispatcher
+    network_mode: host
     hostname: librenms-dispatcher
     cap_add:
       - NET_ADMIN
@@ -89,7 +95,7 @@ services:
       - "TZ=${TZ}"
       - "PUID=${PUID}"
       - "PGID=${PGID}"
-      - "DB_HOST=db"
+      - "DB_HOST=127.0.0.1"
       - "DB_NAME=${MYSQL_DATABASE}"
       - "DB_USER=${MYSQL_USER}"
       - "DB_PASSWORD=${MYSQL_PASSWORD}"
@@ -104,6 +110,7 @@ services:
   syslogng:
     image: /:latest
     container_name: librenms_syslogng
+    network_mode: host
     hostname: librenms-syslogng
     cap_add:
       - NET_ADMIN
@@ -125,7 +132,7 @@ services:
       - "TZ=${TZ}"
       - "PUID=${PUID}"
       - "PGID=${PGID}"
-      - "DB_HOST=db"
+      - "DB_HOST=127.0.0.1"
       - "DB_NAME=${MYSQL_DATABASE}"
       - "DB_USER=${MYSQL_USER}"
       - "DB_PASSWORD=${MYSQL_PASSWORD}"
@@ -136,6 +143,7 @@ services:
   snmptrapd:
     image: /:latest
     container_name: librenms_snmptrapd
+    network_mode: host
     hostname: librenms-snmptrapd
     cap_add:
       - NET_ADMIN
@@ -157,7 +165,7 @@ services:
       - "TZ=${TZ}"
       - "PUID=${PUID}"
       - "PGID=${PGID}"
-      - "DB_HOST=db"
+      - "DB_HOST=127.0.0.1"
       - "DB_NAME=${MYSQL_DATABASE}"
       - "DB_USER=${MYSQL_USER}"
       - "DB_PASSWORD=${MYSQL_PASSWORD}"
```

**Comment by systemcrash** [comment]
> Docker networking is a blocker for this as we need the source IP of the SNMP notification.

IPv6.

File: examples/traefik/docker-compose.yml (hunk 8 above)
Comment (chaitu-tk) [reply examples/traefik/docker-compose.yml:179]: Of course, this means that any conflicts with the host networking have to be resolved, but I guess there is no other way to get SNMP traps.

**Comment by chaitu-tk** [comment]
> > Docker networking is a blocker for this as we need the source IP of the SNMP notification.
> 
> IPv6.

Yes, that is an option as there is no NATing, and libreNMS supports it, but see my comment above able to get it working with `host` network mode in docker with IPV4. We can decide either way.

**Comment by systemcrash** [comment]
### Infrastructure 
If you are doing monitoring, you're setting up dedicated infra, anyway. 

Your example configuration is fairly naïve which means that you run into this NAT problem (and with `host` mode, a potential security hazard which sort-of defeats the point of running containers). iptables on your docker host can help you out here, with e.g. `-j SNAT` or `DNAT` See [this](https://www.engagespark.com/blog/rtp-port-ranges-for-freeswitch-in-docker). 

### Routing
Your problem is one of routing: Route traps to the docker guest destination, not the docker host. You fix this like so:

1. Configure a static subnet for the docker guests, e.g. `172.31.24.0/24` in the compose file
e.g
```
version: '2.1'

networks:
  default:
    driver: bridge
    driver_opts:
      com.docker.network.driver.mtu: 9000
   #enable_ipv6: true
    ipam:
      driver: default
      config:
      - subnet: 172.16.24.0/24
        gateway: 172.16.24.1
```

2. Program a static route at your (default) gateway to `172.31.24.0/24` via your docker host `192.168.1.19`
3. Configure trap senders to send to e.g. `172.31.24.5` directly - this traffic goes to the default GW anyway.

Otherwise, use IPv6.

**Comment by systemcrash** [comment]
Otherwise [`netns` is your friend](https://iximiuz.com/en/posts/container-networking-is-simple/) ;)

**Comment by chaitu-tk** [comment]
> If you are doing monitoring, you're setting up dedicated infra, anyway.

Sure, but not in all cases, and if we merge this PR, we need to be sure it works without any infra expectations.

 > and with host mode, a potential security hazard which sort-of defeats the point of running containers

I understand, it is still helpful as only networking is not containerized, but `bridge` driver offers better security.

> Your problem is one of routing: Route traps to the docker guest destination, not the docker host. 

I can't really try this config right away as I am currently testing on infra which I don't have control (office network), but it should work as the DNAT rule is docker host-based and is not matched. And as you say this also exposes the network to the outside world bypassing docker-proxy, so, this might still pose a security threat.

>  Otherwise netns is your friend ;) (https://iximiuz.com/en/posts/container-networking-is-simple/)

This is exactly what docker `bridge` uses (also mentioned in the article) which doesn't work for us (at least not straight away).

TL;DR, there are three approaches we could follow and if you are worried about security, then `1` is the only option in which case we should enableIPv6 support for snmptrapd in this PR (we could still leave IPv4 in, if someone wants to do 3, but we should document that it might not work OOB).

1. Use IPV6
2. Use IPV4 + Docker `Host` driver + Compose file changes to resolve the docker services
3. Modify the infrastructure to add static routes send traps directly to the container bypassing `docker-proxy`

**Comment by systemcrash** [comment]
Then use:
`iptables -t nat -I PREROUTING  --dst-type LOCAL -p udp -m udp --dport <host listening port> --j DNAT --to-destination <container ip address>:<container port>`

e.g.
`iptables -t nat -I PREROUTING  --dst-type LOCAL -p udp -m udp --dport 162 --j DNAT --to-destination <container ip address>:162`


The DOCKER chain is the DNAT chain.

**Comment by chaitu-tk** [comment]
> Then use:
> `iptables -t nat -I PREROUTING --dst-type LOCAL -p udp -m udp --dport <host listening port> --j DNAT --to-destination <container ip address>:<container port>`
> 
> e.g.
> `iptables -t nat -I PREROUTING --dst-type LOCAL -p udp -m udp --dport 162 --j DNAT --to-destination <container ip address>:162`
> 
> The DOCKER chain is the DNAT chain.

Docker already adds this rule, see below and I guess we are supposed to use DOCKER_USER chain instead of editing DOCKER chain.

```console
Chain DOCKER (2 references)
target     prot opt source               destination         
RETURN     all  --  anywhere             anywhere            
DNAT       tcp  --  anywhere             anywhere             tcp dpt:8000 to:192.168.64.6:8000
DNAT       tcp  --  anywhere             anywhere             tcp dpt:snmp-trap to:192.168.64.7:162
DNAT       udp  --  anywhere             anywhere             udp dpt:snmp-trap to:192.168.64.7:162
```

**Comment by chaitu-tk** [comment]
Oh the command you have shared is for `PREROUTING` CHAIN, not DOCKER, will give it a shot.

File: rootfs/etc/cont-init.d/08-svc-snmptrapd.sh (hunk 7 above)
Comment (glance-) [reply rootfs/etc/cont-init.d/08-svc-snmptrapd.sh:23]: I did it this way because that's the example on https://docs.librenms.org/Extensions/SNMP-Trap-Handler/ , but I agree, saying -m ALL is better.

File: rootfs/etc/snmp/snmptrapd.conf (hunk 6 above)
Comment (glance-) [reply rootfs/etc/snmp/snmptrapd.conf:2]: Yepp. This was just copy paste from https://docs.librenms.org/Extensions/SNMP-Trap-Handler/ , but I've added support for support for substituting COMMUNITYSTRING with $LIBRENMS_SNMP_COMMUNITY.

File: doc/notes/snmptrapd.md (hunk 5 above)
Comment (glance-) [reply doc/notes/snmptrapd.md:9]: SNMP isn't udp only. I'm not intrested in DTLS, so feel free to build that ontop of this in your own PR.

File: examples/traefik/docker-compose.yml (hunk 4 above)
Comment (glance-) [reply examples/traefik/docker-compose.yml:180]: SNMP isn't udp only.

File: examples/rrdcached-server/docker-compose.yml (hunk 3 above)
Comment (glance-) [reply examples/rrdcached-server/docker-compose.yml:172]: SNMP isn't udp only.

File: examples/compose/docker-compose.yml (hunk 2 above)
Comment (glance-) [reply examples/compose/docker-compose.yml:154]: SNMP isn't udp only.

File: Dockerfile (hunk 1 above)
Comment (glance-) [reply Dockerfile:134]: SNMP isn't udp only. I'm not intrested in SNMPv3 or DTLS so feel free to add that ontop of this PR yourself.

Unencrypted SNMP traps over udp and tcp is MVP for me.

**Comment by chaitu-tk** [comment]
> I don't see any problems with getting the ip's propperly mapped for snmp traps in my setup. I'm running docker-ce 20.10.5 and everything works without any issues.

So, you are seeing the SNMP notifications logged to the Event log? In my case `LibreNMS\Snmptrap\Handlers\Trap.php` is not unable to find the IP in the devices list so even though there is a handler registered for that trap it is not called. 

snmptrapd itself shows the notifications properly, but the source IP is mangled.

**Comment by glance-** [comment]
I'm seeing LinkUp.php and LinkDown.php called propperly, without any issues.

**Comment by chaitu-tk** [comment]
Thanks, I have now migrated to the latest version `20.10.6` and see that it is working fine now, I don't remember the previous version, but something has changed in Docker to get this working, sorry for the noise. I am trying to pinpoint to the commit which fixed this, but only found below issues in docker

https://github.com/docker/roadmap/issues/157
https://github.com/moby/moby/issues/15086

**Comment by chaitu-tk** [comment]
LGTM for the initial support we can extend this to SNMPv3, DTLS, TCP etc in the future if we have usecases.

**Comment by systemcrash** [comment]
Hello @murrant - can you ( promote someone who can ) review/approve this PR?

**Review by crazy-max — CHANGES_REQUESTED** [review]
What's the purpose of `data/snmptrapd` folder? Seems not used to me.

**Comment by murrant** [comment]
Looks good to me.

One concern is how do you enable MIBs for snmptrapd... You can set -m ALL, but this can cause a nightmare with broken MIBs and OID collisions.   Additionally, LibreNMS keeps MIBs in sub-directories because putting too many MIBs in one directory can cause the net-snmp MIB parser to have issues.

Changing MIBs also requires a restart of snmptrapd.

**Comment by glance-** [comment]
/data/snmptrapd was a copy paste error when cloning some other service to base the snmptrapd work upon, and just got left there like some appendix.

It's now droped.

**Review by crazy-max — CHANGES_REQUESTED** [review]
Needs rebase to fix CI

**Comment by systemcrash** [comment]
plz check @crazy-max

File: rootfs/etc/cont-init.d/08-svc-snmptrapd.sh [hunk 9]
```diff
@@ -20,6 +20,6 @@ mkdir -p /etc/services.d/snmptrapd
 cat > /etc/services.d/snmptrapd/run <<EOL
 #!/usr/bin/execlineb -P
 with-contenv
-/usr/sbin/snmptrapd -f -m IF-MIB -M /opt/librenms/mibs:/opt/librenms/mibs/cisco
+/usr/sbin/snmptrapd -f -m ALL -M /opt/librenms/mibs:/opt/librenms/mibs/cisco
```
Comment (murrant) [inline rootfs/etc/cont-init.d/08-svc-snmptrapd.sh]: LibreNMS has a lot of MIBs, this might break snmptrapd.
You kind of need away for users to specify this.
It would be cool if there was a way to configure this from LibreNMS, but I have never figured out a way to update snmptrapd at runtime in an os agnostic way.

Comment (glance-) [reply rootfs/etc/cont-init.d/08-svc-snmptrapd.sh]: I don't see a issue with the current way snmptrapd loads mib's, and it works, so I'd suggest if you have the need to configure this you work on top of these patches to add that.

Comment (chaitu-tk) [reply rootfs/etc/cont-init.d/08-svc-snmptrapd.sh]: There is an issue with current approach as it loads only partial MIBs, please note the MIB searching is not recursive, so, you need to specifiy all possible sub-directories as well. This is why it's better to amke it configurable.

Comment (glance-) [reply rootfs/etc/cont-init.d/08-svc-snmptrapd.sh]: I know that mib parsing isn't recursive, but this loads just fine for now, and enough for MVP. Of course it can be improved, but don't let that block this PR.

**Review by crazy-max — APPROVED** [review]
LGTM! PTAL @murrant

**Comment by murrant** [comment]
LGTM, fine with deferring better MIB handling for later, docs should reflect that as it will be a trap for some users to fall in to.
